---
title: "RNAseq practical"
output:
  html_document:
    toc: true
    toc_depth: 2
  pdf_document:
    toc: true
    toc_depth: 2	
  github_document:
    toc: true
    toc_depth: 2	
bibliography: rnaseq_practical.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE, eval=TRUE, cache.lazy = FALSE)

rm(list=ls()) ##clear work space

suppressPackageStartupMessages({ ##to get rid of annoying startup messages
    library(airway)
    library(DESeq2)
    library(vsn)
    library(dplyr)    
    library(ggplot2)
    library(gridExtra)
    library(pheatmap)
    library(RColorBrewer)
    library(genefilter)
})

```

# Introduction #

This practical is based on the BioConductors' *RNA-seq workflow:
gene-level exploratory analysis and differential expression*; a
comprehensive workflow that the describes how to go from FASTQ-files
to perform a differential expression analysis and annotating
results. Here, we will only explore a few steps and focus on the
differential expression analysis. The full workflow is described
[here](https://www.bioconductor.org/help/workflows/rnaseqGene/). The
workflow also appread as a F1000 paper[@Love2015].

## Experimental Data ##

The data used in this workflow is stored in the `airway` package that
summarizes an RNA-seq experiment wherein airway smooth muscle cells
were treated with dexamethasone, a synthetic glucocorticoid steroid
with anti-inflammatory effects [@Himes2014]. Glucocorticoids are used,
for example, by people with asthma to reduce inflammation of the
airways. In the experiment, four primary human airway smooth muscle
cell lines were treated with 1 micromolar dexamethasone for 18
hours. For each of the four cell lines, we have a treated and an
untreated sample. For more description of the experiment see the
[PubMed entry 24926665](https://www.ncbi.nlm.nih.gov/pubmed/24926665)
and for raw data see the GEO entry
[GSE52778](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778).


# Exploratory analysis and visualization #

## Loading and Exploring the data ##

The [airway](http://bioconductor.org/packages/airway/)-package is
available from BioConductor as a data-package and contains both the
gene expression counts as well as metadata on the experiment and
samples. This prepared dataset is what we will use in this practical.

We won't go into the details of how to construct such a dataset or
object but it is good to known that many BioConductor package use
specialized objects to ease various analyses, for example, later we
will see an *DESeqDataSet* which we will use specifically for doing
differential expression analysis using the `DESeq2`-package.

> Use the following code to figure out how many samples and genes are
> in the dataset.

```{r, loading}
library(airway)                         #loading the airway library 
data("airway")                          #loading the airway data
se <- airway                            #for ease of typing
se 
colData(se)
```

> How many million reads are sequenced/aligned to genes for each
> sample?


```{r, totalcounts}
round( colSums(assay(se)) / 1e6, 1 )
```

For the differential expression analysis we will use the
[`DESeq2`](http://bioconductor.org/packages/DESeq2/)-package[@Love2104]. 


## The *DESeqDataSet* object, sample information and the design formula ##

Here we will construct a *DESeqDataSet* from the airway data and add
the design formula containing the covariates on which we will perform
the differential expression analysis.

```{r, deseqdata}
library(DESeq2)
## it is prefered in R that the first level of a factor be the
## reference level (e.g. control, or untreated samples), so we need to
## relevel the dex factor
se$dex <- relevel(se$dex, "untrt")
dds <- DESeqDataSet(se, design = ~ cell + dex) #add formula
dds
```

Our *DESeqDataSet* contains many rows with only zeros, and
additionally many rows with only a few fragments total. In order to
reduce the size of the object, and to increase the speed of our
functions, we can remove the rows that have no or nearly no
information about the amount of gene expression. Here we apply the
most minimal filtering rule: removing rows of the *DESeqDataSet* that
have no counts, or only a single count across all samples. Additional
weighting/filtering to improve power is applied at a later step in the
workflow.

```{r, lowexpressed}
nrow(dds)
dds <- dds[ rowSums(counts(dds)) > 1, ]
nrow(dds)
```

> How many genes have zero counts across all samples?

```{r, zerocounts}
table(rowSums(assays(se)$counts) == 0)
```

## !!!ADVANCED: The rlog and variance stabilizing transformations ##

Many common statistical methods for exploratory analysis of
multidimensional data, for example clustering and principal components
analysis (PCA), work best for data that generally has the same range
of variance at different ranges of the mean values. When the expected
amount of variance is approximately the same across different mean
values, the data is said to be homoskedastic. For RNA-seq counts,
however, the expected variance grows with the mean. For example, if
one performs PCA directly on a matrix of counts or normalized counts
(e.g. correcting for differences in sequencing depth), the resulting
plot typically depends mostly on the genes with highest counts because
they show the largest absolute differences between samples. A simple
and often used strategy to avoid this is to take the logarithm of the
normalized count values plus a pseudocount of 1; however, depending on
the choice of pseudocount, now the genes with the very lowest counts
will contribute a great deal of noise to the resulting plot, because
taking the logarithm of small counts actually inflates their
variance. We can quickly show this property of counts with some
simulated data (here, Poisson counts with a range of lambda from 0.1
to 100). We plot the standard deviation of each row (genes) against
the mean:


```{r, poisson}
lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library(vsn)
msCts <- meanSdPlot(cts, ranks = FALSE, plot=FALSE)
log.cts.one <- log2(cts + 1)
msLog <- meanSdPlot(log.cts.one, ranks = FALSE, plot=FALSE)
library(gridExtra)
grid.arrange(msCts$gg, msLog$gg, nrow=2)
```

The logarithm with a small pseudocount amplifies differences when the
values are close to 0. The low count genes with low signal-to-noise
ratio will overly contribute to sample-sample distances and PCA plots.

For genes with high counts, the rlog and VST will give similar result
to the ordinary log2 transformation of normalized counts. For genes
with lower counts, however, the values are shrunken towards the genes’
averages across all samples. The rlog-transformed or VST data then
becomes approximately homoskedastic, and can be used directly for
computing distances between samples, making PCA plots, or as input to
downstream methods which perform best with homoskedastic data.

```{r, vsnrlog}
library(dplyr)
library(ggplot2)
rld <- rlog(dds, blind = FALSE)
vsd <- vst(dds, blind = FALSE)
dds <- estimateSizeFactors(dds)
df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))  
colnames(df)[1:2] <- c("x", "y")  
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)  
```

Scatterplot of transformed counts from two samples. Shown are
scatterplots using the log2 transform of normalized counts (left),
using the rlog (middle), and using the VST (right). While the rlog is
on roughly the same scale as the log2 counts, the VST has a upward
shift for the smaller values. It is the differences between samples
(deviation from y=x in these scatterplots) which will contribute to
the distance calculations and the PCA plot.

We can see how genes with low counts (bottom left-hand corner) seem to
be excessively variable on the ordinary logarithmic scale, while the
rlog transform and VST compress differences for the low count genes
for which the data provide little information about differential
expression.

## Clustering and PCA ##

A useful first step in an RNA-seq analysis is often to assess overall
similarity between samples: Which samples are similar to each other,
which are different? Does this fit to the expectation from the
experiment’s design?

We use the R function dist to calculate the Euclidean distance between
samples. To ensure we have a roughly equal contribution from all
genes, we use it on the rlog-transformed data. We need to transpose
the matrix of values using t, because the dist function expects the
different samples to be rows of its argument, and different dimensions
(here, genes) to be columns.

```{r, dist}
sampleDists <- dist(t(assay(rld)))
sampleDists
```

We visualize the distances in a heatmap in a figure below, using the
function `pheatmap` from the
[pheatmap](https://cran.r-project.org/web/packages/pheatmap) package.


```{r, heatmap}
library(pheatmap)
library(RColorBrewer)
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( rld$dex, rld$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

> Which samples are more similar to each other?


Another way to visualize sample-to-sample distances is a principal
components analysis (PCA). In this ordination method, the data points
(here, the samples) are projected onto the 2D plane such that they
spread out in the two directions that explain most of the differences
(figure below). The x-axis is the direction that separates the data
points the most. The values of the samples in this direction are
written PC1. The y-axis is a direction (it must be orthogonal to the
first direction) that separates the data the second most. The values
of the samples in this direction are written PC2. The percent of the
total variance that is contained in the direction is printed in the
axis label. Note that these percentages do not add to 100%, because
there are more dimensions that contain the remaining variance
(although each of these remaining dimensions will explain less than
the two that we see).

```{r, pca}
pcaData <- plotPCA(rld, intgroup = c( "dex", "cell"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
```

> Which samples are more similar to each other? What can you say about
> the impact of different cell-lines?


# Differential expression analysis #

## Running the differential expression pipeline ##

As we have already specified an experimental design when we created
the DESeqDataSet, we can run the differential expression pipeline on
the raw counts with a single call to the function DESeq:

```{r, deseq}
dds <- DESeq(dds)
```

This function will print out a message for the various steps it
performs. These are described in more detail in the manual page for
DESeq, which can be accessed by typing ?DESeq. Briefly these are: the
estimation of size factors (controlling for differences in the
sequencing depth of the samples), the estimation of dispersion values
for each gene, and fitting a generalized linear model.

A *DESeqDataSet* is returned that contains all the fitted parameters
within it, and the following section describes how to extract out
results tables of interest from this object.


```{r, results}
res <- results(dds)
res
res[order(res$padj),]
```

Calling results without any arguments will extract the estimated log2
fold changes and p values for the last variable in the design
formula. If there are more than 2 levels for this variable, results
will extract the results table for a comparison of the last level over
the first level. The comparison is printed at the top of the output:
dex trt vs untrt.

The first column, baseMean, is a just the average of the normalized
count values, divided by the size factors, taken over all samples in
the DESeqDataSet. The remaining four columns refer to a specific
contrast, namely the comparison of the trt level over the untrt level
for the factor variable dex. We will find out below how to obtain
other contrasts.

The column log2FoldChange is the effect size estimate. It tells us how
much the gene's expression seems to have changed due to treatment with
dexamethasone in comparison to untreated samples. This value is
reported on a logarithmic scale to base 2: for example, a log2 fold
change of 1.5 means that the gene's expression is increased by a
multiplicative factor of 2^1.5 ~ 2.82.

Of course, this estimate has an uncertainty associated with it, which
is available in the column lfcSE, the standard error estimate for the
log2 fold change estimate. We can also express the uncertainty of a
particular effect size estimate as the result of a statistical
test. The purpose of a test for differential expression is to test
whether the data provides sufficient evidence to conclude that this
value is really different from zero. DESeq2 performs for each gene a
hypothesis test to see whether evidence is sufficient to decide
against the null hypothesis that there is zero effect of the treatment
on the gene and that the observed difference between treatment and
control was merely caused by experimental variability (i.e., the type
of variability that you can expect between different samples in the
same treatment group). As usual in statistics, the result of this test
is reported as a p value, and it is found in the column
pvalue. Remember that a p value indicates the probability that a fold
change as strong as the observed one, or even stronger, would be seen
under the situation described by the null hypothesis.

We can also summarize the results with the following line of code,
which reports some additional information.

```{r, summary}
summary(res)
```

Note that there are many genes with differential expression due to
dexamethasone treatment at the FDR level of 10%. This makes sense, as
the smooth muscle cells of the airway are known to react to
glucocorticoid steroids. However, there are two ways to be more strict
about which set of genes are considered significant:

    lower the false discovery rate threshold (the threshold on padj in
    the results table) 
	
	raise the log2 fold change threshold from 0 using the lfcThreshold
    argument of results

If we lower the false discovery rate threshold, we should also inform
the results() function about it, so that the function can use this
threshold for the optimal independent filtering that it performs:

```{r, resulst2}
res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)
```

If we want to raise the log2 fold change threshold, so that we test
for genes that show more substantial changes due to treatment, we
simply supply a value on the log2 scale. For example, by specifying
lfcThreshold = 1, we test for genes that show significant effects of
treatment on gene counts more than doubling or less than halving,
because 2^1 = 2.

```{r, resulst3}
resLFC1 <- results(dds, lfcThreshold=1)
table(resLFC1$padj < 0.1)
```


> What is the effect of treatment with dexamethasone on the top
> differentially expressed gene?

> How strong is this effect in fold-change comparing treated vs
> untreated?


## !!!ADVANCED: Multiple testing ##

In high-throughput biology, we are careful to not use the p values
directly as evidence against the null, but to correct for multiple
testing. What would happen if we were to simply threshold the p values
at a low value, say 0.05? There are 5676 genes with a p value below
0.05 among the 29391 genes for which the test succeeded in reporting a
p value:

```{r, multipletesting1}
sum(res$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res$pvalue))
```

Now, assume for a moment that the null hypothesis is true for all
genes, i.e., no gene is affected by the treatment with
dexamethasone. Then, by the definition of the p value, we expect up to
5% of the genes to have a p value below 0.05. This amounts to 1470
genes. If we just considered the list of genes with a p value below
0.05 as differentially expressed, this list should therefore be
expected to contain up to 1470 / 5676 = 26% false positives.

`DESeq2` uses the Benjamini-Hochberg (BH) adjustment (Benjamini and
Hochberg 1995) as implemented in the base R `p.adjust` function; in
brief, this method calculates for each gene an adjusted p value that
answers the following question: if one called significant all genes
with an adjusted p value less than or equal to this gene's adjusted p
value threshold, what would be the fraction of false positives (the
false discovery rate, FDR) among them, in the sense of the calculation
outlined above? These values, called the BH-adjusted p values, are
given in the column padj of the res object.

The FDR is a useful statistic for many high-throughput experiments, as
we are often interested in reporting or focusing on a set of
interesting genes, and we would like to put an upper bound on the
percent of false positives in this set.

Hence, if we consider a fraction of 10% false positives acceptable, we
can consider all genes with an adjusted p value below 10% = 0.1 as
significant. How many such genes are there?

```{r, multipletesting2}
sum(res$padj < 0.1, na.rm=TRUE)
```

We subset the results table to these genes and then sort it by the
log2 fold change estimate to get the significant genes with the
strongest down-regulation:

```{r, multipletesting3}
resSig <- subset(res, padj < 0.1)
head(resSig[ order(resSig$log2FoldChange), ])
```
... and with the strongest up-regulation:

```{r, multipletesting4}
head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])
```

## visualizing results ##

A quick way to visualize the counts for a particular gene is to use
the plotCounts function that takes as arguments the DESeqDataSet, a
gene name, and the group over which to plot the counts (figure below).

```{r, plottop}
topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup=c("dex"))
```

In the sample distance heatmap made previously, the dendrogram at the
side shows us a hierarchical clustering of the samples. Such a
clustering can also be performed for the genes. Since the clustering
is only relevant for genes that actually carry a signal, one usually
would only cluster a subset of the most highly variable genes. Here,
for demonstration, let us select the 20 genes with the highest
variance across samples. We will work with the rlog transformed
counts:

The heatmap becomes more interesting if we do not look at absolute
expression strength but rather at the amount by which each gene
deviates in a specific sample from the gene's average across all
samples. Hence, we center each genes' values across samples, and plot
a heatmap (figure below). We provide a data.frame that instructs the
pheatmap function how to label the columns.


```{r, heatmaptop}
library(genefilter)
topVarGenes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 20)
mat  <- assay(rld)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(rld)[, c("cell","dex")])
pheatmap(mat, annotation_col = anno)
```


# Reference #
